mypyc/codegen/emitfunc.py

```diff
@@ -111,6 +125,10 @@ def __init__(self,
         self.literals = emitter.context.literals
         self.rare = False
         self.next_block: Optional[BasicBlock] = None
+        # If not None, the following op is a branch (to allow op merging)
+        self.next_branch: Optional[Branch] = None
```

Comment:
I think that if we're going to go down the route of having the emitter visitor do peephole opts, we should implement it for the general case instead this special case.

One approach that follows this in relying on state in the visitor would be:
 * Stash the current block's ops in a `block_ops` attribute or something
 * And the next loop index into `block_idx`
Then the top level emit loop just uses `visitor.block_idx` as its loop counter and doesn't need any knowledge of when optimizations trigger, and the peephole code in the emitter just looks at `self.block_ops[self.block_idx]` directly and then increments it if the optimization fires.

Another similar approach would be to do that but to have the visitor return `Optional[int]` to indicate skipping instructions.

Reply:
I followed your idea and I think that it looks cleaner now.